extmod/modure.c

```diff
@@ -174,10 +249,184 @@ STATIC mp_obj_t re_split(size_t n_args, const mp_obj_t *args) {
 }
 MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(re_split_obj, 2, 3, re_split);
 
+#if MICROPY_PY_URE_SUB
+// str_to_int ... from objstr.c
+STATIC int str_to_int(const char *str, int *num) {
+    const char *s = str;
+    if ('0' <= *s && *s <= '9') {
+        *num = 0;
+        do {
+            *num = *num * 10 + (*s - '0');
+            s++;
+        } while ('0' <= *s && *s <= '9');
+    }
+    return s - str;
+}
+
+STATIC mp_obj_t ure_exec_sub(mp_obj_re_t *self, mp_obj_t replace, mp_obj_t where, mp_int_t count, mp_int_t flags) {
+    mp_buffer_info_t bufinfo;
+    mp_get_buffer_raise(where, &bufinfo, MP_BUFFER_READ);
+    bool debug = (flags & FLAG_DEBUG);
+    (void)debug;
+    bool ignorecase = (flags & FLAG_IGNORECASE);
+    (void)ignorecase;
+    Subject subj;
+    subj.begin = bufinfo.buf;
+    subj.end = subj.begin + bufinfo.len;
+    int caps_num = (self->re.sub + 1) * 2;
+    int pre_m_l = -1, post_m_l = -1;
+    int num_sub = 0;
+    const char* where_str = subj.begin;
+
+    vstr_t *vstr_return = vstr_new(0);
+    for (;;) {
+        mp_obj_match_t *match = m_new_obj_var(mp_obj_match_t, char*, caps_num);
+        // cast is a workaround for a bug in msvc: it treats const char** as a const pointer instead of a pointer to pointer to const char
+        memset((char*)match->caps, 0, caps_num * sizeof(char*));
+        int res = re1_5_recursiveloopprog(&self->re, &subj, match->caps, caps_num, false);
+
+        // if we didn't have a match, or had an empty match, it's time to stop
+        if (!res || match->caps[0] == match->caps[1]) {
+            break;
+        }
+
+        // process match
+
+        match->base.type = &match_type;
+        match->num_matches = caps_num / 2; // caps_num counts start and end pointers
+        match->str = where;
+
+        const char* replace_str = mp_obj_str_get_str((mp_obj_is_callable(replace) ? mp_call_function_1(replace, MP_OBJ_FROM_PTR(match)) : replace));
+
+        // add pre-match string
+        pre_m_l = (match->caps[0] - subj.begin);
+        if (pre_m_l != -1) {
+            vstr_add_strn(vstr_return, subj.begin, pre_m_l);
+        }
+
+        vstr_t *vstr_repl = vstr_new(0);
+        vstr_add_str(vstr_repl, replace_str);
+        const char *repl_p = vstr_null_terminated_str(vstr_repl);
+        do {
+            const char *group = NULL, *start_group = NULL, *end_group = NULL;
+            int match_no = -1, is_group_number = -1;
+            if (*repl_p == '\\') {
+                start_group = repl_p;
+                group = ++repl_p;
+
+                if (*group != 0 && *group == 'g') {
+                    // search group with syntax "\g<number>"
```

Comment:
@dmazzella is there a need to provide support for both `"\g<number>"` and `"\number"`?  Surely only one format is needed?

Reply:
for reduce code size i will remove "\g<number>" and use only "\number"